readthedocs/proxito/views/serve.py

```diff
@@ -84,23 +84,38 @@ def get(self,
             version_slug=version_slug,
             filename=filename,
         )
-
-        # All public versions can be cached.
         version = final_project.versions.filter(slug=version_slug).first()
-        if (
-            self._is_cache_enabled(final_project)
-            and version and not version.is_private
-        ):
-            self.cache_request = True
 
         log.bind(
             project_slug=final_project.slug,
             subproject_slug=subproject_slug,
             lang_slug=lang_slug,
             version_slug=version_slug,
             filename=filename,
-            cache_request=self.cache_request,
         )
+
+        # Skip serving versions that are not active (return 404). This is to
+        # avoid serving files that we have in the storage, but its associated
+        # version does not exist anymore or it was de-activated.
+        #
+        # Note that we want to serve the page when `version is None` because it
+        # could be a valid URL, like `/` or `` (empty) that does not have a
+        # version associated to it.
+        #
+        # However, if there is a `version_slug` in the URL but there is no
+        # version on the database we want to return 404.
+        if version and not version.active or (not version and version_slug):
```

Comment:
```suggestion
        # However, if there is a `version_slug` in the URL but there is no
        # version on the database we want to return 404.
        if (version and not version.active) or (version_slug and not version):
```

I think we want this, to make it more explicit?

Reply:
Yeah, exactly. That's the problem this logic is trying to solve. I also agree it may be lead to some false positives, but I think we could cover them with test cases. If you have ideas of what other cases I'm missing here, I'm happy to write those test cases as well. Unfortunately, I didn't find a better / more explicit way to write this logic. Let me know.